server/app/services/program/ActiveAndDraftPrograms.java

```diff
@@ -70,30 +81,104 @@ private ImmutableMap<String, ProgramDefinition> mapNameToProgram(
     return mapNameToProgramWithFilter(repository, service, versionModel, Optional.empty());
   }
 
-  private ActiveAndDraftPrograms(VersionRepository repository, Optional<ProgramService> service) {
+  private ActiveAndDraftPrograms(
+      VersionRepository repository,
+      Optional<ProgramService> service,
+      EnumSet<ActiveAndDraftProgramsType> types) {
     VersionModel active = repository.getActiveVersion();
     VersionModel draft = repository.getDraftVersionOrCreate();
     // Note: Building this lookup has N+1 query behavior since a call to getProgramDefinition does
     // an additional database lookup in order to sync the set of questions associated with the
     // program.
+
+    // Active, non-disabled programs.
     ImmutableMap<String, ProgramDefinition> activeNameToProgram =
+        mapNameToProgramWithFilter(repository, service, active, Optional.of(DisplayMode.DISABLED));
+
+    // All active programs (including disabled).
+    ImmutableMap<String, ProgramDefinition> activeNameToProgramAll =
         mapNameToProgram(repository, service, active);
 
+    // Draft, non-disabled programs.
     ImmutableMap<String, ProgramDefinition> draftNameToProgram =
+        mapNameToProgramWithFilter(repository, service, draft, Optional.of(DisplayMode.DISABLED));
+
+    // All draft programs (including disabled).
+    ImmutableMap<String, ProgramDefinition> draftNameToProgramAll =
         mapNameToProgram(repository, service, draft);
 
-    this.activePrograms = activeNameToProgram.values().asList();
-    this.draftPrograms = draftNameToProgram.values().asList();
-    this.versionedByName =
-        Sets.union(activeNameToProgram.keySet(), draftNameToProgram.keySet()).stream()
-            .collect(
-                ImmutableMap.toImmutableMap(
-                    Function.identity(),
-                    programName -> {
-                      return Pair.create(
-                          Optional.ofNullable(activeNameToProgram.get(programName)),
-                          Optional.ofNullable(draftNameToProgram.get(programName)));
-                    }));
+    this.activePrograms =
+        types.equals(allProgramTypes)
+            ? activeNameToProgramAll.values().asList()
+            : activeNameToProgram.values().asList();
+    this.draftPrograms =
+        types.equals(allProgramTypes)
+            ? draftNameToProgramAll.values().asList()
+            : draftNameToProgram.values().asList();
+    if (types.equals(allProgramTypes)) {
+      this.versionedByName =
+          createVersionedByNameMap(activeNameToProgramAll, draftNameToProgramAll);
+    } else if (types.equals(EnumSet.of(ActiveAndDraftProgramsType.DISABLED))) {
```

Comment:
**suggestion:** Can we change this to `types.contains(ActiveAndDraftProgramsType.DISABLED)` here (and similar below)

Reply:
Ok, changes applied.